hedgehog/src/Hedgehog/Internal/Property.hs

```diff
@@ -778,6 +831,38 @@ withRetries :: ShrinkRetries -> Property -> Property
 withRetries n =
   mapConfig $ \config -> config { propertyShrinkRetries = n }
 
+-- | Add a classifier to the test if the predicate is true
+--
+-- @
+--    prop_with_classifier :: Property
+--    prop_with_classifier = property $ do
+--      xs <- forAll $ Gen.list (Range.linear 0 100) Gen.alpha
+--      for_ xs $ \x ->
+--        classify (x == 0) "newborns" $ do
+--        classify (x > 0 && x < 13) "children" $
+--        classify (x > 12 && x < 20) "teens" $
+--          success
+-- @
+classify :: Bool -> String -> PropertyT IO () -> PropertyT IO ()
+classify = cover 0
+
+cover :: Double -> Bool -> String -> PropertyT IO () -> PropertyT IO ()
+cover _ False _ = id
+cover minCoverage True s = PropertyT
+                 . TestT
+                 . mapExceptT (Lazy.mapWriterT addClassification)
+                 . unTest . unPropertyT
+  where
+    addClassification m = do
+      (r, (cl@(Classification cls _), xs)) <- m
+#if __GLASGOW_HASKELL__ != 802
+      -- FIXME GHC 8.2.1 has a bug with referencing `cls` below, thus we need to
+      -- FIXME remove these lines if we're using the 802 series
+      when (Map.member s cls) $ throwError . userError $
+        "classification matched duplicate label: \"" <> s <> "\""
+#endif
```

Comment:
Could you rewrite this without the pattern match to avoid the CPP?

Reply:
I haven't figured out a good way to rephrase that to be perfectly honest. The bug that arises is:

![](https://pbs.twimg.com/media/D13iRjHWkAQ3Y80.png:large)

Which, if I understand it correctly comes from when GHC is trying to figure out how to extract something out of a record. The exact thing that crashes is most likely the `cls` part of: `Map.member s cls`. I'm not sure why the line below the cpp clause works as expected.

Do you have any suggestions?

Or am I misunderstanding you - did you mean just "remove the CPP" stuff?

Reply:
Ah, sorry i thought it was the `<- m` line above, brain fart

hmm that is annoying

Reply:
OK, leave the CPP in, maybe just add the text from that image to the FIXME